multimap/src/main/java/org/infinispan/multimap/impl/Bucket.java

```diff
@@ -76,17 +76,29 @@ public Bucket<V> add(V value) {
       return new Bucket<>(newBucket);
    }
 
-   public Bucket<V> remove(V value) {
+   public Bucket<V> remove(V value, boolean supportsDuplicates) {
       List<V> newBucket = new ArrayList<>(values.size());
       boolean removed = false;
       Iterator<V> it = values.iterator();
-      while (it.hasNext()) {
-         V v = it.next();
-         if (Objects.deepEquals(v, value)) {
-            removed = true;
-            break;
+      if (supportsDuplicates) {
+         int duplicates = (int) values.stream().filter(n -> n.equals(value)).count();
```

Comment:
This code is not correct for a few different reasons. But this code can be simpler.

```java
         while (it.hasNext()){
            V v = it.next();
            if (Objects.deepEquals(v, value)) {
               removed = true;
            } else {
               newBucket.add(v);
            }
         }
```

And the `add remaining values` comment and code snippet can be put in the other else block as this code doesn't need it.